File: svc-position/src/test/java/com/beancounter/position/TestDividends.java
package com.beancounter.position;

import static com.beancounter.common.utils.CurrencyUtils.getCurrency;
import static org.assertj.core.api.Assertions.assertThat;

import com.beancounter.common.model.Asset;
import com.beancounter.common.model.Market;
import com.beancounter.common.model.Portfolio;
import com.beancounter.common.model.Position;
import com.beancounter.common.model.Positions;
import com.beancounter.common.model.Trn;
import com.beancounter.common.model.TrnType;
import com.beancounter.common.utils.AssetUtils;
import com.beancounter.position.service.Accumulator;
import java.math.BigDecimal;
import org.junit.jupiter.api.Test;


class TestDividends {

  @Test
  void is_CashDividendAccumulated() {

    Asset asset = AssetUtils.getAsset("MO", Market.builder()
        .code("ASX")
        .currency(getCurrency("AUD"))
        .build());

    Trn trn = Trn.builder()
        .asset(asset)
        .trnType(TrnType.DIVI)
        .tradeCashRate(new BigDecimal("0.8988"))
        .tradeAmount(new BigDecimal("12.99"))
        .build();

    Accumulator accumulator = new Accumulator();

    Positions positions = new Positions(Portfolio.builder().code("test").build());
    Position position = positions.get(asset);
    accumulator.accumulate(trn, positions.getPortfolio(), position);
    assertThat(position.getMoneyValues(Position.In.TRADE))
        .hasFieldOrPropertyWithValue("dividends", trn.getTradeAmount());

  }
}

File: svc-position/src/test/kotlin/com/beancounter/position/TestDividends.kt
package com.beancounter.position

import com.beancounter.common.model.Currency
import com.beancounter.common.model.Market
import com.beancounter.common.model.Position
import com.beancounter.common.model.Positions
import com.beancounter.common.model.Trn
import com.beancounter.common.model.TrnType
import com.beancounter.common.utils.AssetUtils.Companion.getAsset
import com.beancounter.common.utils.PortfolioUtils.Companion.getPortfolio
import com.beancounter.position.service.Accumulator
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import java.math.BigDecimal

@SpringBootTest(classes = [Accumulator::class])
internal class TestDividends {

    @Autowired
    private lateinit var accumulator: Accumulator

    @Test
    fun is_CashDividendAccumulated() {
        val asx = Market("ASX", Currency("AUD"))
        val asset = getAsset(asx, "MO")
        val trn = Trn(TrnType.DIVI, asset)
        trn.tradeCashRate = BigDecimal("0.8988")
        trn.tradeAmount = BigDecimal("12.99")
        val positions = Positions(getPortfolio("TEST"))
        val position = positions[asset]
        accumulator.accumulate(trn, positions.portfolio, position)
        assertThat(position.getMoneyValues(Position.In.TRADE, asset.market.currency))
            .hasFieldOrPropertyWithValue("dividends", trn.tradeAmount)
    }
}
